crates/router/src/core/payments/helpers.rs

```diff
@@ -5406,6 +5413,539 @@ impl ApplePayData {
+    let expiration_ms = expiration
+        .parse::<i128>()
+        .change_context(errors::GooglePayDecryptionError::InvalidExpirationTime)?;
+    let expiration_time =
+        time::OffsetDateTime::from_unix_timestamp_nanos(expiration_ms * 1_000_000)
+            .change_context(errors::GooglePayDecryptionError::InvalidExpirationTime)?;
+    let now = time::OffsetDateTime::now_utc();
+
+    Ok(expiration_time > now)
+}
+
+// Construct little endian format of u32
+fn get_little_endian_format(number: u32) -> Vec<u8> {
+    number.to_le_bytes().to_vec()
+}
+
+// Filter and parse the root signing keys based on protocol version and expiration time
+fn filter_root_signing_keys(
+    root_keys: Vec<serde_json::Value>,
+) -> CustomResult<Vec<GooglePayRootSigningKey>, errors::GooglePayDecryptionError> {
+    let root_signing_keys = root_keys
+        .into_iter()
+        .map(|key| {
+            serde_json::from_value::<GooglePayRootSigningKey>(key)
+                .change_context(errors::GooglePayDecryptionError::DeserializationFailed)
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let filtered_root_signing_keys = root_signing_keys
+        .iter()
+        .filter(|key| {
+            key.protocol_version == GooglePayProtocolVersion::EcProtocolVersion2
+                && matches!(
+                    check_expiration_date_is_valid(&key.key_expiration).inspect_err(
+                        |err| logger::warn!(
+                            "Failed to check expirattion due to invalid format: {:?}",
+                            err
+                        )
+                    ),
+                    Ok(true)
+                )
+        })
+        .cloned()
+        .collect::<Vec<GooglePayRootSigningKey>>();
+
+    logger::info!(
+        "Filtered {} out of {} root signing keys",
+        filtered_root_signing_keys.len(),
+        root_signing_keys.len()
+    );
+
+    Ok(filtered_root_signing_keys)
+}
+
+impl GooglePayTokenDecryptor {
+    pub fn new(
+        root_keys: masking::Secret<String>,
+        recipient_id: Option<masking::Secret<String>>,
+        private_key: masking::Secret<String>,
+    ) -> CustomResult<Self, errors::GooglePayDecryptionError> {
+        // base64 decode the private key
+        let decoded_key = BASE64_ENGINE
+            .decode(private_key.expose())
+            .change_context(errors::GooglePayDecryptionError::Base64DecodingFailed)?;
+        // create a private key from the decoded key
+        let private_key = PKey::private_key_from_pkcs8(&decoded_key)
+            .change_context(errors::GooglePayDecryptionError::KeyDeserializationFailed)
+            .attach_printable("cannot convert private key from decode_key")?;
+
+        // parse the root signing keys
+        let root_keys_vector: Vec<serde_json::Value> = serde_json::from_str(&root_keys.expose())
+            .change_context(errors::GooglePayDecryptionError::DeserializationFailed)?;
```

Comment:
Can deserialize it to `Vec<GooglePayRootSigningKey>` directly.

Also, prefer to use `parse_struct()`, etc. where possible.